src/Microsoft.AspNet.Authentication.OAuth/OAuthHandler.cs

```diff
@@ -30,131 +29,65 @@ public OAuthHandler(HttpClient backchannel)
 
-        public override async Task<bool> InvokeAsync()
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
         {
-            if (Options.CallbackPath.HasValue && Options.CallbackPath == Request.Path)
-            {
-                return await InvokeReturnPathAsync();
-            }
-            return false;
-        }
+            AuthenticationProperties properties = null;
+            var query = Request.Query;
 
-        public async Task<bool> InvokeReturnPathAsync()
-        {
-            var ticket = await HandleAuthenticateOnceAsync();
-            if (ticket == null)
-            {
-                Logger.LogWarning("Invalid return state, unable to redirect.");
-                Response.StatusCode = 500;
-                return true;
-            }
+            var code = query["code"];
+            var state = query["state"];
 
-            var context = new SigningInContext(Context, ticket)
+            properties = Options.StateDataFormat.Unprotect(state);
+            if (properties == null)
             {
-                SignInScheme = Options.SignInScheme,
-                RedirectUri = ticket.Properties.RedirectUri,
-            };
-            ticket.Properties.RedirectUri = null;
-
-            await Options.Events.SigningIn(context);
+                return AuthenticateResult.Failed("State not found.");
+            }
 
-            if (context.SignInScheme != null && context.Principal != null)
+            // OAuth2 10.12 CSRF
+            if (!ValidateCorrelationId(properties))
             {
-                await Context.Authentication.SignInAsync(context.SignInScheme, context.Principal, context.Properties);
+                return AuthenticateResult.Failed("Correlation failed.");
             }
 
-            if (!context.IsRequestCompleted && context.RedirectUri != null)
+            if (StringValues.IsNullOrEmpty(code))
             {
-                if (context.Principal == null)
-                {
-                    // add a redirect hint that sign-in failed in some way
-                    context.RedirectUri = QueryHelpers.AddQueryString(context.RedirectUri, "error", "access_denied");
-                }
-                Response.Redirect(context.RedirectUri);
-                context.RequestCompleted();
+                return AuthenticateResult.Failed("Code was not found.");
             }
 
-            return context.IsRequestCompleted;
-        }
+            var tokens = await ExchangeCodeAsync(code, BuildRedirectUri(Options.CallbackPath));
 
-        protected override async Task<AuthenticationTicket> HandleAuthenticateAsync()
-        {
-            AuthenticationProperties properties = null;
-            try
+            if (tokens == null || string.IsNullOrEmpty(tokens.AccessToken))
             {
-                var query = Request.Query;
-
-                // TODO: Is this a standard error returned by servers?
-                var value = query["error"];
```

Comment:
Looks like you removed this thing, which may potentially hide why the authentication failed.
Ideally, it would be nice to log `error`, `error_description` and `error_uri` (when populated by the authorization server).

(see https://tools.ietf.org/html/rfc6749#section-4.1.2.1 for more info about `error`, `error_description` and `error_uri`)